---
title: "Functions for lab"
author: "Sarah Hu"
date: "3/5/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# RNA extractions

Qiagen RNA extraction protocol.
```{r}
rlt_beta <- function(sample_num, buffer_ml){
  total_needed <- sample_num * buffer_ml
  cat("You need", total_needed, "mLs of RLT buffer.\n")
  beta <- (total_needed * 10)
  cat("Prep RLT by adding", beta, "µl of beta-mercaptoethanol to", total_needed, "mL of RLT.")
}
rlt_beta(5, 2)
# 1.5 * 8
# 5+8
```
## EtOH
```{r}
# CV=CV EtOH
# Assumes you start with 100% EtOH
etoh <- function(ml, percent){
  vol_etoh <- (ml*(percent/100))/1
  vol_h20 <- ml - vol_etoh
  cat("Prep", percent, "% EtOH by adding", vol_h20, "ml water +", vol_etoh, "ml of 100% ethanol.")
}
etoh(500, 70)
```


```{r}
DNase <- function(sample_num, error){
  RDD <- 70 * (sample_num + error)
  dnaseI <- 10 * (sample_num + error)
  total <- (RDD + dnaseI)
  cat("Combine", RDD, "µl of RDD buffer, with", dnaseI, "µl of DNase I enzyme.\n")
  cat("This is enough master mix for", sample_num, "samples, and totals to", 
      total, "µls in volume. Add 80 µl to each RNase column membrane.")
}
DNase(8, 0.01)
# 488/6
# 550/81
# 480.8/80
```


# RNA quantification
How much 1X TE would you need total?
```{r}
TE_total <- function(sample_num){
  stdcurve_num <- 8
  totalwells <- sample_num + stdcurve_num
  cat("With 8 wells designated for the standard curve, and", sample_num, "samples, account for", totalwells, "wells on the plate. And 1999 µl for making Ribogreen mixture.\n")
  ribo <- 2000 #2ml of Ribogreen
  ul_total <- ((totalwells * 50) + ribo)
  cat("Make", ul_total, "µl of 1X TE, or", (ul_total/1000), "mls.\n\n")
  tomake_mls <- ceiling(ul_total/1000)
  ml2 <- (tomake_mls/2)
  ml1 <- (tomake_mls/1)
  cat("Option 1: Make", ml1, "total solutions of 1X TE in 1ml aliquots:\n50µl of 20X TE + 950µl water. Repeat",ml1,"times.\n\n\n")
  cat("Option2: Make", ml2, "total solutions of 1X TE in 2ml aliquots:\n100µl of 20X TE + 1900µl water. Repeat",ml2,"times.\n")
}
TE_total(16)
# 2000/20
```

# cDNA calculation
This is based on the iScript Select cDNA synthesis step from BioRad.
```{r}
cDNA_mix <- function(sample_num, err, desired_conc){
  cat("With", sample_num, "samples for cDNA. Make this master mix:\n")
  sample_num_werr <- (sample_num + (sample_num * err))
  rxn_mix5x <- (4 * sample_num_werr)
  rt <- (1 * sample_num_werr)
  primer <- (2 * sample_num_werr)
  cat("5X IScript rev rxn mix:", rxn_mix5x, "µl\n")
  cat("iScript RT enzyme:", rt, "µl\n")
  cat("Random primers:", primer, "µl\n\n")
  total_vol <- rxn_mix5x + rt + primer
  cat("Total vol of master mix is", total_vol, "µl.\n",
      "Add 7µl to each tube. Add RNA and water to 20µl.")
  # CV=CV
}

cDNA_mix(7.1, 0.0) #0.01 = 1% error

4*17.5
1*17.5
2*17.5
```

# PCR reaction calculation
This is based on the Q5 High-Fidelity 2X Master Mix from BioRad.
Assumes that final volume of reaction is 25µl and the input concentration of primers is 10µM.
```{r}
PCR_master_mix <- function(sample_num, err, desired_conc, desired_vol){
  sample_num_werr <- (sample_num + (sample_num * err))
  rxn_mixQ5 <- (12.5 * sample_num_werr)
  primer <- (1.25 * sample_num_werr)
  vol_water <- (sample_num_werr * (10-desired_vol))
  input_template_conc <- ((desired_conc/25) * 25)/desired_vol
  cat("Solve for input DNA template concentration:\n")
  cat("(C)(V) =", (desired_conc/25) * 25, "ng\n")
  cat("To add", desired_vol, "µl to all PCR reactions, make sure input DNA template has a concentration of", input_template_conc, "µl.\n\n")
  cat("To prep for", sample_num, "PCR reactions. Make this master mix:\n")
  cat("Q5 Hi-Fi 2X Mix:", rxn_mixQ5, "µl\n")
  cat("Forward primer (10µM):", primer, "µl\n")
  cat("Reverse primer (10µM):", primer, "µl\n")
  cat("Nuclease-free water:", vol_water,"µl\n")
  cat("*Add above water to master mix if all DNA inputs are normalized to add", desired_vol, "µl of template.\n")
  each_tube <- (12.5 + 1.25 + 1.25 + (10-desired_vol))
  cat("Prep each tube by adding", each_tube,"µl of master mix, and", desired_vol, "µl of DNA template.\n\n")
  cat("Otherwise, add 15µl of master mix to each tube. Then add 10µl total of water and input template DNA so total reaction volume is 25µl\n\n")
  # CV=CV
}
# arguments: sample_num, err, desired_conc, desired_vol
PCR_master_mix(4, 0.01, 1, 5) #0.01 = 1% error
```

Prep gel and PCR products to run.
```{r}
# perc_agarose - for 1% input 0.01
# function automatically adds 2 samples for error

run_gel <- function(perc_agarose, total_ml, sample_num){
  cat("Make sure to prep a gel with combs for", sample_num+2, "samples with bp ladder.\n")
  grams <- (perc_agarose * total_ml)
  cat("To prep the agarose, add", grams, "grams to", total_ml, "mls of 1X TAE. Microwave to dissolve into solution. Let cool, add SYBR safe, and pour with combs.\n\n")
  water <- 2 * (sample_num + 2)
  dye <- 1 * (sample_num + 2)
  cat("Make a solution of", water, "µls of water and", dye, "µls of loading dye. Add 3 µls on top of parafilm (label with sample numbers) and mix ontop of parafilm with 2µls of your PCR amplicon.\n Add 5µl of this to each well in the gel.")
}
run_gel(0.01, 75, 25)
```

# AMPure bead clean-up
```{r}
bead_clean <- function(sample_num, etoh_perc){
  total_ul <- sample_num * 400
  in_ml <- total_ul/1000
  cat("You need", in_ml, "mls of EtOH to wash all samples.\n")
  # (cv)=(cv)
  vol_etoh <- etoh_perc * total_ul
  cat("Add", vol_etoh, "µls to", (total_ul-vol_etoh), "µl of water, or\n",
      "add", (vol_etoh)/1000, "mls to", (total_ul-vol_etoh)/1000, "ml of water.")
}
bead_clean(5, 0.8)
```


# Sequencing calculations

Was sequencing budget given in Gb of base pairs?
```{r}
Gb_to_readpair <- function(Gb, ended, length){
  # Gb == billion bps per lane
  # ended * length = total bps
  readpair <- (((Gb * 1E9)/(ended * length)))/1E9
  cat("With", Gb ,"billion bps and", length, paste(ifelse(ended == 1, "SE", "PE")), "sequencing, run will give",readpair, "billion", paste(ifelse(ended == 1, "SE", "PE")), "reads per lane.")
}

Gb_to_readpair(2400, 2, 150)
```

Coverage depth refers to the average number of sequencing reads that align to, or "cover," each base in your sequenced sample. The Lander/Waterman equation1 is a method for calculating coverage (C) based on your read length (L), number of reads (N), and haploid genome length (G): C = LN / G
```{r}
# ended 1 = SE and 2 = PE
# length = length of read (e.g., 100 or 150) in bps
# genome_length = effective genome size (total) in Gb
# coverage = number of times you want to sequence over the genome
coverage_depth_est <- function(length, ended, genome_length, coverage){
  # coverage = (read length * number of reads)/genome length
  est_num <- (length*ended) * genome_length * coverage
  cat("To get", coverage, "x coverage, you need at least", est_num, "reads.")
}
# If I want 100X coverage, how many reads do I need to get?
coverage_depth_est(150, 2, 200, 100)
```


```{r}
# million_user = million read pairs a user wants (if you want 50M read pairs coverage, use '50')
# ended 1 = SE and 2 = PE
# length = length of read (e.g., 100 or 150)
# sample_num = samples you want to sequence
# billion_seq = billion read pairs available in 1 lane of sequencing (1 billion, use '1')
# lanes = number of lanes

seq_calc <- function(million_user, ended, length, sample_num, billion_seq, lanes, machine){
  cat("Sequencing target:\n")
  cat("I want", million_user, "Million read pairs for", sample_num, "total samples.\n")
  sample_bps <- ((million_user * 1E6) * (ended * length)) # Bps per sample user wants
  cat("Therefore, the total bps desired per sample =", sample_bps, "bps\n")
  total_readpairs <- ((million_user * 1E6) * sample_num)
  total_bps <- (sample_bps * sample_num)
  cat("For", sample_num, "samples, that totals to", total_bps, "bps need to be sequenced, to get my desired coverage. That's", total_readpairs," total read pairs.\n\n")
  ##
  cat("Available sequencing:\n")
  avail <- (billion_seq * 1E9) # Read pairs available
  cat(machine, "sequencing can give me", avail, "reads per lane (", billion_seq,"billion read pairs ).\n","Which comes out to", avail*lanes, "total reads across", lanes, "lanes.\n\n")
  samples_avail <- (billion_seq * 1E9)/(million_user * 1E6)
  ##
  cat("Possible sequence analysis:\n")
  cat("I can then target", samples_avail, "samples, to get approximately", million_user, "Million reads per sample per lane.\n")
  cat("Across", lanes, "total lanes, I can get", (samples_avail * lanes), "samples.\n")
  ## $
  need_lanes <- (million_user * 1E6)/(billion_seq * 1E9)
  cat("Since you have", sample_num, "samples to get sequenced, you need to budget for at least", sample_num/samples_avail, "( or", round((sample_num/samples_avail),0),")",machine, "lanes.\n\n")
  ##
  cat("Build budget to cover library prep and QC for",sample_num,"samples and", round((sample_num/samples_avail),0), machine, "lanes.")
}
# seq_calc(80, 2, 150, 20, 2, 4, "NovaSeq")
seq_calc(100, 2, 150, 20, 2, 4, "NovaSeq")
seq_calc(200, 2, 150, 58, 2, 4, "NovaSeq")
```

# Qubit math
```{r}
n <- 10
# Make sure to include x2 for standards
```

```{r}
buffer <- 199
reagent <- 1

cat("Add", (buffer*(n)),"µl of buffer and", (reagent*n), "µl of reagent for relevant kit.\n\n",
    "For 2 standards, use 190µl of working solution and 10µl of standards.\n\n", "For each sample, use 198µl of working solution + 2µl of sample.")
```


# NEB Ultra II Directional RNA library prep

Indicate number of samples for library prep.
```{r}
n <- 16
```


## First Strand Reaction Buffer and Random Primer Mix
```{r}
NEB_Next_1st <- 8
Ran_pri <- 2
h2o <- 10

total <- NEB_Next_1st + Ran_pri + h2o


cat("Main mix:\n", ((NEB_Next_1st)*(n)), "µl of NEB Next 1st Strand Syn Rxn Buffer\n",
    (Ran_pri*n), "µl of Random primers\n",
    (h2o*n), "Nuclease-free water\n\n", "Use this working solution to elute mRNA from beads, add 11.5µl to each samples. ")
```

## mRNA isolation bead step
```{r}
oligo_beads <- 20
RNA_bind_buff <- 100
total <- oligo_beads+RNA_bind_buff

cat("Main mix:\n", 
    (oligo_beads*n), "µl Oligo dT Beads d(T)25\n",
    (RNA_bind_buff*n), "µl RNA Binding Buffer x2\n",
    "Add above to 1.5ml tube.\nWash beads by pipetting up and down x6.\n Place on magnet.\n For RNA Binding Buffer wash, add", (100*n), "µl to the 1.5 ml tube and mix.\n\nFor final step, add", (50*n), "µl of RNA Binding Buffer to the 1.5 ml tube.\n Outcome is beads that have been washed.")

```

## Second Strand cDNA synthesis
```{r}
NEB_Next_2nd_reagent <- 8
NEB_Next_2_ndEnzyme <- 4
h2o <- 48

cat("Main mix:\n", ((NEB_Next_2nd_reagent)*(n)), "µl of NEBNext 2nd Strand Synthesis Rxn Buffer with dUTP Mix\n",
    (NEB_Next_2_ndEnzyme*n), "µl of NEBNext 2nd Strand synthesis Enzyme mix\n",
    (h2o*n), "µl of Nuclease-free water.\n",
    "ON ICE, add 60 µl to all 1st strand synthesis products (which should all be 20 µl at this point too)")

# NEB_Next_2_ndEnzyme + NEB_Next_2nd_reagent+ h2o
```

## Adapter ligation step
```{r}
NEB_Ligation_enhancer <- 1
NEB_Ligation_MasterMix <- 30

cat("First, make up adapter dilution with Adapter Dilution BUffer and NEBNext Adapter. Keep this on ice. For 100-fold, add 1µl of adapter to 99µl of dilution buffer.\n\n","Mix Ligation enhancer and master mix ahead of time:\n", ((NEB_Ligation_enhancer)*(n)), "µl of NEBNext NEBNext Ligation Adapter\n",
    (NEB_Ligation_MasterMix*n), "µl of NEBNext Ligation Master Mix\n",
    "To each sample (prepped DNA), add 2.5 µl of adapter dilution. Then, add 31 µl of ligation mix to each reaction")
```

# Ovation Nugen DNA library kit

```{r}
# Number of samples
n <- 12
```

For whole protocol, there are several 70% EtOH washes.
```{r}
total_etoh <- (200 * 8)

cat("Will need a total of:\n", ((total_etoh)*(n)), "µl of 70% EtOH\n",
    ((total_etoh)*(n))/1000, "ml")
```

Table 3 in protocol - End Repair Master Mix
```{r}
ER1 <- 3.5
ER2 <- 0.5
ER3 <- 1.0
nPLUS <- 0.2 + n
cat("End repair master mix:\n", 
    ((nPLUS)*(ER1)), "µl of ER1 - End Repair Buffer Mix\n",
    ((nPLUS)*(ER2)), "µl of ER2 - End Repair Enzyme Mix\n",
    ((nPLUS)*(ER3)), "µl of ER3 - End Repair Enhancer\n",
        (((nPLUS)*(ER1))+((nPLUS)*(ER2))+((nPLUS)*(ER3))), "µl total vol\n\nAdd 5µl to each reaction")
```
Table 4 - Ligation
```{r}
D1 <- 1.5
L1 <- 6.0
L3 <- 1.5
nPLUS <- 0.2 + n
cat("Ligation master mix:\n", 
    ((nPLUS)*(D1)), "µl of D1 - water\n",
    ((nPLUS)*(L1)), "µl of L1 - Ligation buffer Mix\n",
    ((nPLUS)*(L3)), "µl of L3 - Ligation enzyme mix\n",
        (((nPLUS)*(D1))+((nPLUS)*(L1))+((nPLUS)*(L3))), "µl total vol\n\nAdd 9µl to each reaction")

```

Library amplification
```{r}
P1 <- 12.75
P2 <- 1.25
P3 <- 1.0
nPLUS <- 0.2 + n
cat("Ligation master mix:\n", 
    ((nPLUS)*(P1)), "µl of P1 - Amp Buffer Mix\n",
    ((nPLUS)*(P2)), "µl of P2 - Amp primer mix\n",
    ((nPLUS)*(P3)), "µl of P3 - Amp enzyme mix\n",
        (((nPLUS)*(P1))+((nPLUS)*(P2))+((nPLUS)*(P3))), "µl total vol\n\nAdd 15 µl to each reaction")
```


# Sequencing library prep
Calculate molarity nM
## Molarity calculation
```{r}
#ng/µl * 1,000,000 / (insert size) * 660 = nM

# conc must be in ng/ul

molarity_lib <- function(insert, conc){
  nM <- ((conc)*1000000)/((insert)*660)
  cat(nM, "nM")
}

# if concentration is in pg/µl multiply by 0.001

# molarity_lib(438, (0.001*2053.73))


molarity_lib(424.87, (0.001*925.53))
# sample 3, 1ng of mRNA
# molarity_lib(349.8, (0.001*2152.72))

molarity_lib(425, 3.49)
```

